### DISCUSSION TITLE
How do I disable the *Pyright* unused expression diagnostic only for *molten*?

### DISCUSSION THREAD

**Discussion by stevenxxiu** [benlubas/molten-nvim#146] [Q&A] [answered]
Thanks for the plugin! I'm still in the stages of setting it up.

The *Pyright* diagnostic is quite useful by itself, and I don't want to disable it for *Python* files. How do I only disable it for *molten*?

I thought about using some per-buffer *lspconfig* for *Pyright*, but I think it uses the same config everywhere.

I couldn't find anything in *otter* either.

**Comment by benlubas** [comment]
This is a good question...

Somewhere not too long ago I saw a reddit comment talking about custom filters for LSP diagnostics. I think that post was in the context of lua LS and ignoring `_unused_variables` that had a leading underscore. (or maybe it was the same question but for pyright.)

Let me try to find that.

**Reply by benlubas** [reply]
Here is one thread with a potentially relevant answer: https://www.reddit.com/r/neovim/comments/108tjy0/nvimlspconfig_how_to_disable_hints_for_unused/

Here is the relevant answer from `u/Resident_One_7912`:

```lua
local function filter(arr, func)
    -- Filter in place
    -- https://stackoverflow.com/questions/49709998/how-to-filter-a-lua-array-inplace
    local new_index = 1
    local size_orig = #arr
    for old_index, v in ipairs(arr) do
        if func(v, old_index) then
            arr[new_index] = v
            new_index = new_index + 1
        end
    end
    for i = new_index, size_orig do arr[i] = nil end
end

local function pyright_accessed_filter(diagnostic)
    -- Allow kwargs to be unused, sometimes you want many functions to take the
    -- same arguments but you don't use all the arguments in all the functions,
    -- so kwargs is used to suck up all the extras
    -- if diagnostic.message == '"kwargs" is not accessed' then
    -- 	return false
    -- end
    --
    -- Allow variables starting with an underscore
    -- if string.match(diagnostic.message, '"_.+" is not accessed') then
    -- 	return false
    -- end


    -- For all messages "is not accessed"
    if string.match(diagnostic.message, '".+" is not accessed') then
        return false
    end

    return true
end

local function custom_on_publish_diagnostics(a, params, client_id, c, config)
    filter(params.diagnostics, pyright_accessed_filter)
    vim.lsp.diagnostic.on_publish_diagnostics(a, params, client_id, c, config)
end

-- IN LSP on_attach FUNCTION
local lsp_opts = {
    on_attach = function(client, bufnr)
       ...


        if client.name == "pyright" then
            vim.keymap.set('n', '<Leader>o', '<CMD>PyrightOrganizeImports<CR>')
            vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
                custom_on_publish_diagnostics, {})

        end
    end,
}
```


You can probably use this and modify the `pyright_accessed_filter` to instead filter out unused variable diagnostics for certain filetypes that you use molten in.

If you want to disable for all buffers when molten is initialized for that buffer, you can use `require('molten.status').kernels() ~= ""` to determine when there's a kernel attached to the current buffer.


Let me know what you come up with, and I can add it to the docs.

**Reply by stevenxxiu** [reply, marked as answer]
Thanks for the template there.

I first tried adding this to `cmp.lua`, but turns out `on_attach` isn't called there. I think *otter* does its own thing.

So instead I attached to an event, `LspAttach`, that *otter* does report.

To see if it's a *molten* buffer, I just tested the file name. I think this is more robust, as there might not be kernels that are active.

Putting the following in `otter.lua` solves the problem:

```lua
local function filter(arr, func)
  -- Filter in place <https://stackoverflow.com/questions/49709998/how-to-filter-a-lua-array-inplace>
  local new_index = 1
  local size_orig = #arr
  for old_index, v in ipairs(arr) do
    if func(v, old_index) then
      arr[new_index] = v
      new_index = new_index + 1
    end
  end
  for i = new_index, size_orig do
    arr[i] = nil
  end
end

local function pyright_molten_filter(diagnostic)
  -- Allow unused expressions, as they often come at the end
  if diagnostic.message == 'Expression value is unused' then
    return false
  end
  return true
end

local function user_on_publish_diagnostics(a, params, client_id, c, config)
  if string.match(params.uri, '%.otter.py$') then
    filter(params.diagnostics, pyright_molten_filter)
  end
  vim.lsp.diagnostic.on_publish_diagnostics(a, params, client_id, c, config)
end

return {
  'jmbuhr/otter.nvim',
  event = 'VeryLazy',
  config = function()
    vim.api.nvim_create_autocmd('LspAttach', {
      callback = function(args)
        local client = vim.lsp.get_client_by_id(args.data.client_id)
        if client.name == 'pyright' then
          vim.lsp.handlers['textDocument/publishDiagnostics'] = vim.lsp.with(user_on_publish_diagnostics, {})
        end
      end,
    })
  end,
}
```

Feel free to test it.

**Reply by benlubas** [reply]
Oh of course molten isn't going to attached to the otter buffer which is where the diagnostics are actually reported.

I'm glad that worked out for you!

**Comment by mariduv** [comment]
I was just looking to filter out these same warnings, but noted `vim.lsp.with` is deprecated in 0.11 while reading through the above answer, and while trying to simplify down, switched to adding this filter in the neovim 0.11+ lsp config style.

This is what I have in `.config/nvim/after/lsp/basedpyright.lua` (or `pyright.lua` if you're using that):

```lua
--- @type vim.lsp.Config
return {
  handlers = {
    ['textDocument/publishDiagnostics'] = function(_, params, ctx)
      -- otter.nvim is an adapter to run regular lsps on the code cells of
      -- journals. it makes hidden buffers of all the code of each language.
      -- filter out warnings for stuff that is normal (an "unused" last
      -- expression is still displayed)
      if string.match(params.uri, "%.otter.py") then
        params.diagnostics = vim.tbl_filter(function(it)
          return it.code ~= "reportUnusedCallResult" and
            it.code ~= "reportUnusedExpression"
        end, params.diagnostics)
      end

      return vim.lsp.diagnostic.on_publish_diagnostics(_, params, ctx)
    end,
  }
}
```
